pennylane/tape/interfaces/autograd.py

```diff
@@ -107,26 +97,34 @@ def _update_trainable_params(self):
         self.trainable_params = trainable_params
         self._all_parameter_values = params
 
-    def get_parameters(self, trainable_only=True):  # pylint: disable=missing-function-docstring
-        self._update_trainable_params()
-        params = self._all_parameter_values
+    def get_parameters(
+        self, trainable_only=True, return_arraybox=False
+    ):  # pylint: disable=missing-function-docstring
+        params = []
+        iterator = self.trainable_params if trainable_only else self._par_info
 
-        if trainable_only:
-            params = [
-                p
-                for idx, p in enumerate(self._all_parameter_values)
-                if idx in self.trainable_params
-            ]
+        for p_idx in iterator:
+            op = self._par_info[p_idx]["op"]
+            op_idx = self._par_info[p_idx]["p_idx"]
+            params.append(op.data[op_idx])
 
-        return autograd.builtins.list(params)
+        return params if return_arraybox else autograd.builtins.list(params)
 
     @autograd.extend.primitive
     def _execute(self, params, device):
         # unwrap all NumPy scalar arrays to Python literals
         params = [p.item() if p.shape == tuple() else p for p in params]
         params = autograd.builtins.tuple(params)
 
+        # unwrap constant parameters
+        all_params_unwrapped = [
+            p.numpy() if isinstance(p, np.tensor) else p for p in self._all_parameter_values
+        ]
+
+        # evaluate the tape
+        self.set_parameters(all_params_unwrapped, trainable_only=False)
         res = self.execute_device(params, device=device)
+        self.set_parameters(self._all_parameter_values, trainable_only=False)

```

Comment:
Good point. The second call to `set_parameters`, where the unwrapped parameters are replaced with the wrapped ones, is more to avoid surprising the user. That is, undoing any side effects that tape execution had.

This won't be required if using a QNode, but if using a tape, you might notice the parameter types had changed before/after executing the tape on a device 🤔

Reply:
Ideally, after this PR is merged we can re-visit and remove all the `set_parameter()` logic in all three interfaces.